### PR TITLE
[C-Api] tensor data handle and API defines @open sesame 7/5 17:55

### PR DESCRIPTION
**Pull request by jaeyun-jung**
As we discussed, tensors metadata in c-api changed to the handles for meta/data.

1. alloc and destory functions are added for tensors meta/data handle.
2. refactors all functions and testcases with meta/data handle.
3. update api description.

Signed-off-by: Jaeyun Jung <jy1210.jung@samsung.com>

### CODE REVIEW

File: api/capi/src/nnstreamer-capi-util.c [hunk 1]
```diff
@@ -424,5 +425,22 @@ int
 ml_util_check_nnfw (ml_nnfw_e nnfw, ml_nnfw_hw_e hw)
 {
   /** @todo fill this function */
```
Comment (again4you) [inline api/capi/src/nnstreamer-capi-util.c]: What about removing this `@todo` item?

File: api/capi/src/nnstreamer-capi-single.c [hunk 2]
```diff
@@ -123,14 +124,57 @@ ml_single_open (ml_single_h * single, const char *model_path,
...
+  switch (nnfw) {
+    case ML_NNFW_UNKNOWN:
+      if (g_str_has_suffix (path_down, ".tflite")) {
+        ml_logi ("The given model [%s] is supposed a tensorflow-lite model.", model_path);
+        nnfw = ML_NNFW_TENSORFLOW_LITE;
+      } else if (g_str_has_suffix (path_down, ".pb")) {
+        ml_logi ("The given model [%s] is supposed a tensorflow model.", model_path);
+        nnfw = ML_NNFW_TENSORFLOW;
+      } else {
+        ml_loge ("The given model [%s] has unknown extension.", model_path);
```
Comment (again4you) [inline api/capi/src/nnstreamer-capi-single.c]: `g_free (path_down);` is required.

File: api/capi/src/nnstreamer-capi-single.c [hunk 3]
```diff
@@ -123,14 +124,57 @@ ml_single_open (ml_single_h * single, const char *model_path,
...
+        return ML_ERROR_INVALID_PARAMETER;
+      }
+      break;
+    case ML_NNFW_CUSTOM_FILTER:
+      if (!g_str_has_suffix (path_down, ".so")) {
+        ml_loge ("The given model [%s] has invalid extension.", model_path);
```
Comment (again4you) [inline api/capi/src/nnstreamer-capi-single.c]: ditto.

File: api/capi/src/nnstreamer-capi-single.c [hunk 4]
```diff
@@ -123,14 +124,57 @@ ml_single_open (ml_single_h * single, const char *model_path,
...
+        return ML_ERROR_INVALID_PARAMETER;
+      }
+      break;
+    case ML_NNFW_TENSORFLOW_LITE:
+      if (!g_str_has_suffix (path_down, ".tflite")) {
+        ml_loge ("The given model [%s] has invalid extension.", model_path);
```
Comment (again4you) [inline api/capi/src/nnstreamer-capi-single.c]: ditto.

File: api/capi/src/nnstreamer-capi-single.c [hunk 5]
```diff
@@ -123,14 +124,57 @@ ml_single_open (ml_single_h * single, const char *model_path,
...
+        return ML_ERROR_INVALID_PARAMETER;
+      }
+      break;
+    case ML_NNFW_TENSORFLOW:
+      if (!g_str_has_suffix (path_down, ".pb")) {
+        ml_loge ("The given model [%s] has invalid extension.", model_path);
```
Comment (again4you) [inline api/capi/src/nnstreamer-capi-single.c]: ditto.

File: api/capi/src/nnstreamer-capi-util.c (hunk 1 above)
Comment (jaeyun-jung) [reply api/capi/src/nnstreamer-capi-util.c]: it was removed.

File: api/capi/src/nnstreamer-capi-single.c (hunk 2 above)
Comment (jaeyun-jung) [reply api/capi/src/nnstreamer-capi-single.c]: updated to free path string

File: api/capi/include/nnstreamer.h [hunk 6]
```diff
@@ -235,12 +224,12 @@ typedef struct {
  * @detail If an application wants to accept data outputs of an NNStreamer stream, use this callback to get data from the stream. Note that the buffer may be deallocated after the return and this is synchronously called. Thus, if you need the data afterwards, copy the data to another buffer and return fast. Do not hold too much time in the callback. It is recommended to use very small tensors at sinks.
  * @since_tizen 5.5
  * @remarks The @a data can be used only in the callback. To use outside, make a copy.
- * @remarks The @a info can be used only in the callback. To use outside, make a copy.
- * @param[in] data The contents of the tensor output (a single frame. tensor/tensors). Number of tensors is determined by data->num_tensors. Note that max num_tensors is 16 (ML_TENSOR_SIZE_LIMIT).
- * @param[in] info The cardinality, dimension, and type of given tensor/tensors.
+ * @remarks The @a info_h can be used only in the callback. To use outside, make a copy.
```
Comment (again4you) [inline api/capi/include/nnstreamer.h]: * info_h -> info

> ../nnstreamer/api/capi/include/nnstreamer.h:227: Question:
  @a is used to refer to parameters, but the 'info_h' parameter was not found in this function. Are you sure this is correct?

File: api/capi/include/nnstreamer.h [hunk 7]
```diff
@@ -531,28 +618,28 @@ void ml_util_free_tensors_info (ml_tensors_info_s *info);
...
  * @return @c Tensors data pointer allocated. Null if error. Caller is responsible to free the allocated data with ml_util_free_tensors_data().
  * @retval NULL There is an error. Call ml_util_get_last_error() to get specific error code.
  */
-ml_tensors_data_s *ml_util_allocate_tensors_data (const ml_tensors_info_s *info);
+ml_tensors_data_s *ml_util_allocate_tensors_data (const ml_tensors_info_h info);
 
 /**
  * @brief Checks the availability of the given execution environments.
  * @since_tizen 5.5
+ * @details If the function returns an error, @a available is not changed.
```
Comment (again4you) [inline api/capi/include/nnstreamer.h]: Please fix it as below. 

```cpp
* @details If the function returns an error, @a available is not changed.
* @since_tizen 5.5
```

For your information, correct tag order is as below.
* Correct tag order:
  @deprecated (mandatory if applicable)
  @platform or @partner
    If @privlevel is 'platform' or 'partner', respectively.
    Should not be used in other cases.
  @brief (mandatory)
  @details
  @since_tizen (mandatory)
  @privlevel
  @privilege (privlevel/privilege can be repeated if necessary)
  @remarks
  @param
  @return (mandatory if return type is not void)
  @retval
  @exception
  @pre
  @post
  @see

File: api/capi/include/nnstreamer.h (hunk 6 above)
Comment (jaeyun-jung) [reply api/capi/include/nnstreamer.h]: It was changed to info.

File: api/capi/include/nnstreamer.h (hunk 7 above)
Comment (jaeyun-jung) [reply api/capi/include/nnstreamer.h]: I updated this.

**Comment by jaeyun-jung** [comment]
I added new commit to add buffer handle. Please review the changes again.

**Comment by jaeyun-jung** [comment]
@pmilkowski @again4you 
I added a new PR for your comment. Please review the changes again.

**Comment by again4you** [comment]
* Comment 14, 15, 16, 17, 18, 19, 20, 25 are done.
* Comment 24 is keep!

**Comment by pmilkowski** [comment]
I checked the issues I mentioned above and they all seem to be fixed. Thank you!

**Pull request merged**